internal/controllers/rollout/controller.go

```diff
@@ -25,114 +22,81 @@ type controller struct {
 	cooldown time.Duration
 }
 
-// NewController re-synthesizes compositions when their synthesizer has changed while honoring a cooldown period.
 func NewController(mgr ctrl.Manager, cooldown time.Duration) error {
 	c := &controller{
 		client:   mgr.GetClient(),
 		cooldown: cooldown,
 	}
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&apiv1.Synthesizer{}).
-		Watches(&apiv1.Composition{}, newCompositionHandler()).
-		// TODO: Filter some events?
-		WithLogConstructor(manager.NewLogConstructor(mgr, "synthesizerRolloutController")).
+		Named("rolloutController").
+		Watches(&apiv1.Composition{}, manager.SingleEventHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "rolloutController")).
 		Complete(c)
 }
 
 func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := logr.FromContextOrDiscard(ctx)
 
-	syn := &apiv1.Synthesizer{}
-	err := c.client.Get(ctx, req.NamespacedName, syn)
-	if err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("gettting synthesizer: %w", err))
-	}
-	logger = logger.WithValues("synthesizerName", syn.Name, "synthesizerNamespace", syn.Namespace, "synthesizerGeneration", syn.Generation)
-
-	compList := &apiv1.CompositionList{}
-	err = c.client.List(ctx, compList, client.MatchingFields{
-		manager.IdxCompositionsBySynthesizer: syn.Name,
-	})
+	comps := &apiv1.CompositionList{}
+	err := c.client.List(ctx, comps)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("listing compositions: %w", err)
 	}
 
-	var latestRollout time.Time
-	for _, comp := range compList.Items {
-		if comp.Status.CurrentSynthesis != nil && comp.Status.PreviousSynthesis != nil && comp.Status.CurrentSynthesis.Synthesized != nil && comp.Status.CurrentSynthesis.Synthesized.Time.After(latestRollout) {
-			latestRollout = comp.Status.CurrentSynthesis.Synthesized.Time
+	// Find the current cooldown period, wait for the next if needed
+	// Block on any active deferred syntheses with less than 2 attempts to avoid high concurrency
+	var lastDeferredSynthesisCompletion *metav1.Time
+	for _, comp := range comps.Items {
+		current := comp.Status.CurrentSynthesis
+		if current == nil || !current.Deferred {
+			continue
+		}
+		if current.Synthesized == nil && current.Attempts < 2 {
```

Comment:
Are we relying solely on exponential backoff to limit concurrency if we have a bunch of compositons being retried?

Reply:
Yeah for now. I still think we may need to add pod concurrency limits at some point though